src/transformers/retrieval_rag.py

```diff
@@ -0,0 +1,535 @@
+        ids = [[int(self.index_id_to_db_id[doc_id]) for doc_id in doc_ids] for doc_ids in docs_ids]
+        return torch.tensor(ids), torch.tensor(vectors)
+
+
+class HFIndex(Index):
+    """
+    A wrapper around an instance of :class:`~datasets.Datasets`. If ``index_path`` is set to ``None``,
+    we load the pre-computed index available with the :class:`~datasets.arrow_dataset.Dataset`, otherwise, we load the index from the indicated path on disk.
+
+    Args:
+        dataset (:obj:`str`, optional, defaults to ``wiki_dpr``):
+            A datatset identifier of the indexed dataset on HuggingFace AWS bucket (list all available datasets and ids with ``datasets.list_datasets()``).
+        dataset_split (:obj:`str`, optional, defaults to ``train``)
+            Which split of the ``dataset`` to load.
+        index_name (:obj:`str`, optional, defaults to ``train``)
+            The index_name of the index associated with the ``dataset``. The index loaded from ``index_path`` will be saved under this name.
+        index_path (:obj:`str`, optional, defaults to ``None``)
+            The path to the serialized faiss index on disk.
+    """
+
+    def __init__(
+        self,
+        dataset,
+        dataset_split,
+        index_name,
+        index_path,
+        use_dummy_dataset,
+    ):
+        super().__init__()
+        self.dataset = dataset
+        self.dataset_split = dataset_split
+        self.index_name = index_name
+        self.index_path = index_path
+        self._index_initialize = False
+        self.use_dummy_dataset = use_dummy_dataset
+
+        logger.info("Loading passages from {}".format(self.dataset))
+        self.index = load_dataset(
+            self.dataset, with_index=False, split=self.dataset_split, dummy=self.use_dummy_dataset
+        )
+
+    def is_initialized(self):
+        return self._index_initialize
+
+    def init_index(self):
+        if self.index_path is not None:
+            logger.info("Loading index from {}".format(self.index_path))
+            self.index.load_faiss_index(index_name=self.index_name, file=self.index_path)
+        else:
+            logger.info("Loading index from {}".format(self.dataset + " with index name " + self.index_name))
+            self.index = load_dataset(
+                self.dataset,
+                with_embeddings=True,
+                with_index=True,
+                split=self.dataset_split,
+                index_name=self.index_name,
+                dummy=self.use_dummy_dataset,
+            )
+        self._index_initialize = True
+
+    def get_doc_dicts(self, doc_ids):
+        return [self.index[doc_ids[i].tolist()] for i in range(doc_ids.shape[0])]
+
+    def get_top_docs(self, question_hidden_states, n_docs=5):
+        _, docs = self.index.get_nearest_examples_batch("embeddings", question_hidden_states, n_docs)
+        ids = [[int(i) for i in doc["id"]] for doc in docs]
```

Comment:
`doc["id"]` is not necessarily the id of `doc` in the `Dataset`. For example with the dummy dataset, `self.index[0]["i"]` returns `'1'`

The brute force `transformers`-only solution would be to return:
```
ids = [[self.index['id'].index(i) for i in doc["id"]] for doc in docs]
```

But I think the `datasets` `get_nearest_examples_batch` function should really return the actual dataset index list @lhoestq

Reply:
Also @ola13 here

Reply:
The ids in the tsv file from the DPR repo start at 1.
Therefore for the HfIndex the 'id' column values start at 1 (id of the first sample is 1)

Also I think the LegacyIndex takes this shift by on into account as well since `LegacyIndex.index_id_to_db_id` is also shifted by one

Reply:
Gotcha. I'd still rather return the actual dataset indexes without relying on the `"id"` field since I'm not sure whether it will be present (or what its semantics will be) in future supporting datasets. Would there be a reason to use the field instead of the index?

Reply:
I'd rather use the index as well.
Maybe we could also remove the `index_id_to_db_id` if it just corresponds to the index shifted by one. Is that correct @ola13 ?

Reply:
Hey I replicated the logic from the original DPR repo (here and around here https://github.com/facebookresearch/DPR/blob/39247e2fa538e5f05f83e0482dff699e1f8415d5/dpr/indexer/faiss_indexers.py#L88) - my bad for not checking it's just shifting by one, if that's the case we can definitely simplify, however I haven't really checked that.

Overall `LegacyIndex` is hand crafted to serve this specific dataset we used in our training (for results reproducibility) and I think going forward any other indides should be built on top of `HFIndex` - so I think it's just fine to check the behavior for the files we have and refactor based on evidence from there.